### PR TITLE
Refactor ContextMenuWrapper

### PR DESCRIPTION
**Pull request by benwolfram**
**Problem:**
Modernizing this file to show it some love in preparation of work to improve context menus and upgrade react-contexify, ie. keyboard navigation.

**Fix:**
Updating to modern react hooks and functional components.

Only note, is that I chose to remove the memoization (formerly `shouldComponentUpdate`, replace with `memo` in one of my earlier commits). I don't see a performance impact from this and it's better to not pre-emptively memoize. Typescript wasn't able to properly infer the generic types when wrapping with memo (see the commit related to removing memo).

s/o to @Rheeseyb for the help

**Manual Tests:**
I hereby swear that:

- [x] I opened a hydrogen project and it loaded
- [x] I could navigate to various routes in Preview mode

Relates to https://github.com/concrete-utopia/utopia/issues/5643

### CODE REVIEW

File: editor/src/components/context-menu-wrapper.tsx [hunk 1]
```diff
@@ -18,142 +27,125 @@ import type { WindowPoint } from '../core/shared/math-utils'
...
+}
+
+interface SimpleItem<T> {
+  item: Item<T>
+  type: 'simple'
+}
+
+type MenuItem<T = null> = Submenu<T> | SimpleItem<T> | null
+
+export type Item<T = null> = ContextMenuItem<T> | null
```
Comment (benwolfram) [inline editor/src/components/context-menu-wrapper.tsx:43]: Low key, I don't love how I've defaulted to null here, but it works.

Haven't dealt much with generics, used to preferring to have explicit typing — so open to any suggestions. Particularly I hadn't encountered component's having types to them, ie. `export class MomentumContextMenu<T> extends ReactComponent<ContextMenuProps<T>>` where this first `T` is coming from and how to seamlessly translate to functional component refactor.

File: editor/src/components/context-menu-wrapper.tsx [hunk 2]
```diff
@@ -18,206 +27,195 @@ import type { WindowPoint } from '../core/shared/math-utils'
...
+
+export type ContextMenuData<T> = Item<T>[] | null
+
 export interface ContextMenuWrapperProps<T> {
-  id: string
-  items: Array<ContextMenuItem<T>>
-  dispatch?: EditorDispatch
-  data: T
-  renderTag?: string
+  children?: ReactNode
```
Comment (Rheeseyb) [inline editor/src/components/context-menu-wrapper.tsx:48]: Normally we'd use `React.PropsWithChildren` but this is fine tbh

**Review by Rheeseyb — APPROVED** [review]

File: editor/src/components/context-menu-wrapper.tsx [hunk 3]
```diff
@@ -18,206 +27,195 @@ import type { WindowPoint } from '../core/shared/math-utils'
...
+              if (!(triggerEvent instanceof MouseEvent)) return null
+              return windowPoint({ x: triggerEvent.clientX, y: triggerEvent.clientY })
+            })()
+            item?.action(getData(), dispatch, rightClickCoordinate, event)
+            contextMenu.hideAll()
+          }}
+          hidden={isHidden(item)}
+          style={{
+            height: item?.isSeparator ? 9 : 28,
+            display: 'flex',
```
Comment (maltenuhn) [inline editor/src/components/context-menu-wrapper.tsx:140]: Maybe too soon / out of scope for this refactor: but I wonder if we'd be better off with a `grid` here, or even a variable for the list-item layout that we want (just label, label with accesor, icon+ label, icon+label+accessor etc).

File: editor/src/components/context-menu-wrapper.tsx [hunk 4]
```diff
@@ -18,206 +27,195 @@ import type { WindowPoint } from '../core/shared/math-utils'
...
+            alignItems: 'center',
+            borderRadius: 4,
```
Comment (maltenuhn) [inline editor/src/components/context-menu-wrapper.tsx:142]: some of these should come from `UtopiaTheme` (same for height)

File: editor/src/components/context-menu-wrapper.tsx [hunk 5]
```diff
@@ -18,206 +27,195 @@ import type { WindowPoint } from '../core/shared/math-utils'
...
+      style={style}
+      onMouseDown={stopPropagation}
+      onMouseUp={stopPropagation}
+      onClick={stopPropagation}
+      ref={forwardRef}
+    >
+      <MenuProvider id={id} itemsLength={items.length} key={`${id}-provider`} style={providerStyle}>
+        {children}
+      </MenuProvider>
+      <MomentumContextMenu dispatch={dispatch} getData={getData} id={id} items={items} key={id} />
```
Comment (maltenuhn) [inline editor/src/components/context-menu-wrapper.tsx:216]: small thing, but can we rename this to just `ContextMenu` (or `UtopiaContextMenu` as much as that feels redundant)? Momentum was an old product name we used in the mid-1990s

File: editor/src/components/context-menu-wrapper.tsx [hunk 6]
```diff
@@ -18,206 +27,195 @@ import type { WindowPoint } from '../core/shared/math-utils'
...
+                  {item.label}
+                </span>
+              }
+              arrow={<Icons.ExpansionArrowRightWhite style={{ marginLeft: 8 }} />}
+            >
+              {item.items.map(renderItem)}
+            </SubmenuComponent>
+          )
+        } else {
+          if (item === null) return null
```
Comment (gbalint) [inline editor/src/components/context-menu-wrapper.tsx:175]: Minor styling thing, but we avoid using single line if statements

**Review by gbalint — APPROVED** [review]

File: editor/src/components/context-menu-wrapper.tsx [hunk 7]
```diff
@@ -2,7 +2,16 @@
 /** @jsx jsx */
 import { jsx } from '@emotion/react'
 import fastDeepEquals from 'fast-deep-equal'
-import React, { Component as ReactComponent } from 'react'
+import React, {
+  type CSSProperties,
```
Comment (liady) [inline editor/src/components/context-menu-wrapper.tsx:6]: minor - we need to decide at some point whether we're going with `React.<Something>` or `import { Something } from 'react'`

File: editor/src/components/context-menu-wrapper.tsx [hunk 8]
```diff
@@ -238,75 +236,79 @@ export const InspectorRowHoverCSS = {
...
+  id,
+  itemsLength,
+  localXHack_KILLME,
+  style,
+}: MenuProviderProps) => {
+  const { show } = useContextMenu({ id })
+
+  const onContextMenu = useCallback(
+    (event: MouseEvent<HTMLDivElement>) => {
+      if (itemsLength <= 0) return
```
Comment (liady) [inline editor/src/components/context-menu-wrapper.tsx:297]: single-line here as well :)

File: editor/src/components/context-menu-wrapper.tsx [hunk 9]
```diff
@@ -238,75 +236,79 @@ export const InspectorRowHoverCSS = {
...
+
+      show(
+        event,
+        localXHack_KILLME === 'local-x-coord-KILLME'
```
Comment (liady) [inline editor/src/components/context-menu-wrapper.tsx:301]: again, minor - but maybe extract it to 
```ts
  const options = localXHack_KILLME === 'local-x-coord-KILLME' 
      ? { position: { x: event.nativeEvent.offsetX, y: event.nativeEvent.pageY } } 
      : undefined

  show(event, options)
```

File: editor/src/components/context-menu-wrapper.tsx [hunk 10]
```diff
@@ -18,206 +27,195 @@ import type { WindowPoint } from '../core/shared/math-utils'
...
 
-  onShown = () => {
-    document.body.classList.add(BodyMenuOpenClass)
-  }
-  onHidden = () => {
-    document.body.classList.remove(BodyMenuOpenClass)
-  }
+  const renderItem = useCallback(
+    (item: Item<T>, index: number) => {
+      return (
```
Comment (liady) [inline editor/src/components/context-menu-wrapper.tsx:123]: how easy/hard will it be to extract it to its own component instead of a callback?

**Review by liady — APPROVED** [review]
lgtm, left some code style comments

File: editor/src/components/context-menu-wrapper.tsx (hunk 5 above)
Comment (benwolfram) [reply editor/src/components/context-menu-wrapper.tsx:216]: Sounds good, I was curious what Momentum meant tbh, didn't really understand!

File: editor/src/components/context-menu-wrapper.tsx (hunk 7 above)
Comment (benwolfram) [reply editor/src/components/context-menu-wrapper.tsx:6]: Roger that, feels weird to me, but I gotta learn all the preferences here, so grill me on them all! Thanks.

File: editor/src/components/context-menu-wrapper.tsx (hunk 10 above)
Comment (benwolfram) [reply editor/src/components/context-menu-wrapper.tsx:123]: I probably should do this.. I'll update!

File: editor/src/components/context-menu-wrapper.tsx (hunk 5 above)
Comment (benwolfram) [reply editor/src/components/context-menu-wrapper.tsx]: Done in it's own commit

File: editor/src/components/context-menu-wrapper.tsx (hunk 10 above)
Comment (benwolfram) [reply editor/src/components/context-menu-wrapper.tsx:113]: So... this felt easy and like a no brainer for organization. But I feel like there are just hidden traps waiting for me in this component 😂. I spent all afternoon trying to get it and making it a new component broke the 'Paste' functionality and tests. So I've reverted that.

If you think it's important, maybe we can follow up later.

**Pull request merged**